Polish Rspotify

ramsayleung opened the pull request:
Clean up Rspotify, take some points of https://github.com/ramsayleung/rspotify/issues/127 into consideration.
+ [x] ~~remove convert_map_to_string and convert_string_to_map in `util.rs`, replaced them with `Reqwest's` `query` method and `Url` struct.~~
+ [x] Many functions like Spotify::tracks take in a Vec<&str>/Vec<String>/&[String]/&[&str], when they should take in an impl IntoIterator<Item = &str>.
+ [x] Split single `senum.rs` file into a module named `enums` and move it into `model` module.
+ [x] reexport struct in `model` module to allow users to write `rspotify::model::FullAlbum` instead of the overly verbose `rspotify::model::album::FullAlbum`.
+ [x] Rename CurrentlyPlaybackContext to CurrentPlaybackContext.
+ [x] Change release_date_precision to enum.

marioortizmanero commented on src/client.rs:
```diff
@@ -244,7 +232,7 @@ impl Spotify {
     pub async fn tracks(
         &self,
-        track_ids: Vec<&str>,
+        track_ids: impl IntoIterator<Item = &str>,
```
One thing I've realized is the inconsistency with `impl IntoIterator<Item = &str>`. The changelog indicates that this is the new function signature for iterable parameters, but sometimes you use `impl IntoIterator<Item = String>` or `impl IntoIterator<Item = &String>`. I would go with the `&str` solution to avoid unnecessary allocations, and only use that type in this module.

ramsayleung replied:
> One thing I've realized is the inconsistency with impl IntoIterator<Item = &str>. The changelog indicates that this is the new function signature for iterable parameters

Yes, I have noticed the `inconsistency` with `impl IntoIterator<Item = &str>`, but it's caused by the original design, not the introduction of `impl IntoIterator<Item = &str>`. 

> but sometimes you use impl IntoIterator<Item = String> or impl IntoIterator<Item = &String>

Because some functions use `&[String]` and some functions use `Vec<&str>`, If I change them to `impl IntoIterator<Item = &String>`, it will break up the backward compatibility.

> I would go with the &str solution to avoid unnecessary allocations, and only use that type in this module.

I have thinked about this solution, it's fine to break up backward compatibility at this release.

marioortizmanero replied:
I would say it's OK to break backward compatibility with this. We're going to change lots of other things anyway.

marioortizmanero commented on src/oauth2.rs:
```diff
@@ -379,13 +378,14 @@ impl SpotifyOAuth {
+        Url::parse_with_params(AUTHORIZE_URL, payload.clone())
+            .unwrap_or_else(|_| {
+                panic!(
+                    "Fail to generate authorize_url, url: {}, payload: {:?}",
```
This should be "Failed", not "Fail"

marioortizmanero commented on CHANGELOG.md:
```diff
@@ -11,7 +11,10 @@
   + Cleanup, reduced repetitive code and boilerplate internally in several places ([#117](https://github.com/ramsayleung/rspotify/pull/117), [#113](https://github.com/ramsayleung/rspotify/pull/113), [#107](https://github.com/ramsayleung/rspotify/pull/107), [#106](https://github.com/ramsayleung/rspotify/pull/106)).
-- Updated dependencies to the latest versions, integrated Dependabot to keep track of them ([#105](https://github.com/ramsayleung/rspotify/pull/105), [#111](https://github.com/ramsayleung/rspotify/pull/111)).
+  + Updated dependencies to the latest versions, integrated Dependabot to keep track of them ([#105](https://github.com/ramsayleung/rspotify/pull/105), [#111](https://github.com/ramsayleung/rspotify/pull/111)).
+  + Endpoints take `Vec<String>/[String]` as parameter have changed to `impl IntoIterator<Item = &str>`, which is backward compatibility.
```
IMO this should be:

```
Endpoints that take `Vec<String>/[String]` as a parameter have changed to `impl IntoIterator<Item = &str>`.
```

And just be left outside the "Breaking changes" section.

ramsayleung commented:
@marioortizmanero I think this PR is ready to review, as for the rest of issues mention in #127 which are mostly about Model, I prefer to create a new PR, otherwise this PR will be a huge one which is difficult to review.

marioortizmanero commented:
The thing is that I'm doing *tons* of changes at #129, specially related to the endpoints, so the merge conflict is going to be huge if we merge this right now. Can you remove the changes you've done to the endpoints and OAuth to avoid this? We can take care of them after it's merged.

Very sorry for the inconvenience. I thought it was going to require less changes but I was forced to even rewrite the OAuth part. The endpoints work differently now as well, so I had to change them as well.

ramsayleung commented:
>  #129 is almost ready as well, can you start taking a look in there? If you can help me to update the docs and tests it'd be super useful.

Yeah, Sure. I am gonna to review thie PR, and collaborate to update docs and tests :)

ramsayleung commented on src/blocking/client.rs:
```diff
@@ -180,15 +179,14 @@ impl Spotify {
-                builder.json(json)
-            } else {
-                builder
+            let builder = match method {
```
@marioortizmanero Check this function for more details.

marioortizmanero commented:
The diff is all messed up for this PR. Is there a way to remove my commits from the PR we just merged so that I can review this?

marioortizmanero commented on src/client.rs:
```diff
@@ -11,29 +11,29 @@ use thiserror::Error;
-use super::model::recommend::Recommendations;
-use super::model::search::SearchResult;
-use super::model::show::{
+use super::model::CUDResult;
```
I think this is one of the few acceptable moments where you can use `use super::model::*` and `use super::enums::*`. The client should be using all of them anyway, and it cleans up the header a lot. What do you think? That, or merging all the `use super::model` into a huge single import. I prefer the first option, though.

ramsayleung replied:
Yep, it's a acceptable choice to use `use super::model::*`, or we have to import a lot of model files individually.

marioortizmanero commented on CHANGELOG.md:
```diff
@@ -12,8 +12,10 @@ If we missed any change or there's something you'd like to discuss about this ve
-- Updated dependencies to the latest versions, integrated Dependabot to keep track of them ([#105](https://github.com/ramsayleung/rspotify/pull/105), [#111](https://github.com/ramsayleung/rspotify/pull/111)).
-
+  + Updated dependencies to the latest versions, integrated Dependabot to keep track of them ([#105](https://github.com/ramsayleung/rspotify/pull/105), [#111](https://github.com/ramsayleung/rspotify/pull/111)).
++ Remove `convert_map_to_str` and `convert_str_to_map` from both `util.rs` and `blocking/util.rs`, replacing them with `reqwest`'s `query` and `Url::Parse`. Remove crate `percent-encoding`
```
This can be removed now.

marioortizmanero commented on CHANGELOG.md:
```diff
@@ -12,8 +12,10 @@ If we missed any change or there's something you'd like to discuss about this ve
-
+  + Updated dependencies to the latest versions, integrated Dependabot to keep track of them ([#105](https://github.com/ramsayleung/rspotify/pull/105), [#111](https://github.com/ramsayleung/rspotify/pull/111)).
++ Remove `convert_map_to_str` and `convert_str_to_map` from both `util.rs` and `blocking/util.rs`, replacing them with `reqwest`'s `query` and `Url::Parse`. Remove crate `percent-encoding`
++ Remove `generate_random_string` and `datetime_to_timestamp` from `blocking/util.rs`, using `generate_random_string` and `datetime_to_timestamp` from `util.rs` instead.
```
This can be removed now.

marioortizmanero commented on CHANGELOG.md:
```diff
@@ -38,6 +40,33 @@ If we missed any change or there's something you'd like to discuss about this ve
+	- `user_unfollow_users`
+	- `audios_features`
+- Reexport `model` module to allow user to write `rspotify::model::FullAlbum` instead of  `rspotify::model::FullAlbum`.
+- Split single `senum.rs` file into a seperate module named `enums` (which is more appropriate compared with `senum`) with three files `country.rs`, `types.rs`, `misc.rs`, check [enums mod.rs file](./src/enums/mod.rs) for details.
```
Typo: "separate", not "seperate"

marioortizmanero commented on CHANGELOG.md:
```diff
@@ -38,6 +40,33 @@ If we missed any change or there's something you'd like to discuss about this ve
+	- `user_follow_users`
+	- `user_unfollow_users`
+	- `audios_features`
+- Reexport `model` module to allow user to write `rspotify::model::FullAlbum` instead of  `rspotify::model::FullAlbum`.
```
I think you got confused here

ramsayleung replied:
Sorry for not getting your point.

marioortizmanero replied:
I just don't understand this line. Both import paths are the same.

ramsayleung replied:
Ooops, I get it now. I'll fix it.

marioortizmanero commented on CHANGELOG.md:
```diff
@@ -38,6 +40,33 @@ If we missed any change or there's something you'd like to discuss about this ve
+	- `user_artist_check_follow`
+	- `user_follow_users`
+	- `user_unfollow_users`
+	- `audios_features`
```
Now that I realize, `audios_features` should be renamed to `audio_features`. Should we do that in this PR or leave it for later?

ramsayleung replied:
I could fix it in this PR.

marioortizmanero replied:
sure, that's probably easier

ramsayleung replied:
Well, it turns out that it's not a problem, we have both `audio_features` and `audios_features` endpoint:

+ audio_features: Get audio features for a track
+ audios_features: Get Audio Features for Several Tracks

marioortizmanero replied:
Ah ok. How about `track` instead of `audio`? "audios" seems off to me, perhaps `track_features` and `tracks_features` would be more fitting

ramsayleung replied:
Yep, it's a more suitable name, I will rename `audio_features`, `audios_features` and `audio_analysis` to `track_features`, `tracks_features` and `track_analysis`.

marioortizmanero commented on examples/with_refresh_token.rs:
```diff
@@ -43,7 +43,7 @@ async fn do_things(spotify: Spotify) {
 
     spotify
-        .user_unfollow_artists(&artists)
+        .user_unfollow_artists(artists.clone())
```
This `clone` is not necessary.

marioortizmanero commented on src/client.rs:
```diff
@@ -1309,10 +1339,13 @@ impl Spotify {
         let url = format!(
             "me/following/contains?type=artist&ids={}",
-            artsit_ids.join(",")
+            artsit_ids.into_iter().collect::<Vec<&str>>().join(",")
```
The `collect` can usually be simplified to `collect::<Vec<_>>()`.

marioortizmanero commented on src/client.rs:
```diff
@@ -1325,8 +1358,14 @@ impl Spotify {
+    ) -> ClientResult<()> {
+        let url = format!(
+            "me/following?type=user&ids={}",
+            user_ids.into_iter().collect::<Vec<&str>>().join(",")
```
Same here, the collect can be simplified.

marioortizmanero commented on src/client.rs:
```diff
@@ -1339,8 +1378,14 @@ impl Spotify {
+    ) -> ClientResult<()> {
+        let url = format!(
+            "me/following?type=user&ids={}",
+            user_ids.into_iter().collect::<Vec<&str>>().join(",")
```
The collect can be simplified.

marioortizmanero commented on src/client.rs:
```diff
@@ -1894,8 +1939,8 @@ impl Spotify {
-    pub async fn save_shows(&self, ids: Vec<String>) -> ClientResult<()> {
-        let joined_ids = ids.join(",");
+    pub async fn save_shows<'a>(&self, ids: impl IntoIterator<Item = &'a str>) -> ClientResult<()> {
+        let joined_ids = ids.into_iter().collect::<Vec<&str>>().join(",");
```
The collect can be simplified.

marioortizmanero commented on src/enums/types.rs:
```diff
@@ -0,0 +1,254 @@
+    kind: ErrorKind,
+}
+
+impl Error {
```
Couldn't this Error be removed and merge the ErrorKind with ClientError? Something like `ParseEnum` would be fine IMO.

marioortizmanero commented on src/client.rs:
```diff
@@ -1952,14 +1997,17 @@ impl Spotify {
-        params.insert("ids".to_owned(), ids.join(","));
+        params.insert(
+            "ids".to_owned(),
+            ids.into_iter().collect::<Vec<&str>>().join(","),
```
The collect can be simplified.

marioortizmanero commented on src/client.rs:
```diff
@@ -2031,13 +2079,16 @@ impl Spotify {
-        params.insert("ids".to_owned(), ids.join(","));
+        params.insert(
+            "ids".to_owned(),
+            ids.into_iter().collect::<Vec<&str>>().join(","),
```
The collect can be simplified.

marioortizmanero commented on src/client.rs:
```diff
@@ -2052,9 +2103,15 @@ impl Spotify {
-        params.insert("ids".to_owned(), ids.join(","));
+        params.insert(
+            "ids".to_owned(),
+            ids.into_iter().collect::<Vec<&str>>().join(","),
```
The collect can be simplified.

marioortizmanero commented on src/client.rs:
```diff
@@ -2068,12 +2125,12 @@ impl Spotify {
         market: Option<Country>,
     ) -> ClientResult<()> {
-        let joined_ids = ids.join(",");
+        let joined_ids = ids.into_iter().collect::<Vec<&str>>().join(",");
```
The collect can be simplified.

marioortizmanero commented on src/client.rs:
```diff
@@ -1294,8 +1318,14 @@ impl Spotify {
+    ) -> ClientResult<()> {
+        let url = format!(
+            "me/following?type=artist&ids={}",
+            artist_ids.into_iter().collect::<Vec<&str>>().join(",")
```
This collect can be simplified

marioortizmanero commented on src/client.rs:
```diff
@@ -1280,8 +1298,14 @@ impl Spotify {
+    ) -> ClientResult<()> {
+        let url = format!(
+            "me/following?type=artist&ids={}",
+            artist_ids.into_iter().collect::<Vec<&str>>().join(",")
```
This collect can be simplified

marioortizmanero reviewed (commented):
I've left some comments about problems I found when looking at the code. There's two things I want to point out:

* Why is `enum` a separate module? Can't this be included in the `model` module as well, as after all, they are part of the models? Or am I just wrong here?
* The conversions to/from enums are incredibly repetitive. Can we use a custom macro or something like [`strum`](https://docs.rs/strum/0.19.5/strum/) instead?

marioortizmanero commented on examples/with_refresh_token.rs:
```diff
@@ -43,7 +43,7 @@ async fn do_things(spotify: Spotify) {
 
     spotify
-        .user_unfollow_artists(artists.clone())
+        .user_unfollow_artists(artists)
```
Ah my bad. It seems I was wrong. You do need the `clone`

ramsayleung commented:
>  Why is enum a separate module? Can't this be included in the model module as well, as after all, they are part of the models? Or am I just wrong here?

Actually, the `enum` is a single file name `senum.rs`(means `enums`, a bad joke), but with more and more `enum` added, the `senum` file grown up to be a huge one, but I separated it into a module, with multiple 

> The conversions to/from enums are incredibly repetitive. Can we use a custom macro or something like strum instead?

Of course we could, I could take time to figure out how to create a decent macro.

marioortizmanero commented:
> Actually, the enum was a single file named senum.rs(which means enums, a bad joke) before, but with more and more enums added, the senum file grown up to be a huge one, but I separated it into a module, with multiple files.

But why is it not merged with the `model` module? It kinda seems like the same to me, but I'm not sure if I understand it correctly.

ramsayleung commented:
> But why is it not merged with the model module? It kinda seems like the same to me, but I'm not sure if I understand it correctly.

Yes, we could, perhaps it's more consistent to move `enums` module into `model` module.

Another problem I want to discuss is that which import path do you prefer,  `rspotify::model::FullAlbum` or `rspotify::FullAlbum`. 

The former one is longger, and explicit, the later one is shorter and shorter. I personnally perfer to the former one.

marioortizmanero commented:
I think it's better to keep that in the `model` module as well.

marioortizmanero commented on src/enums/types.rs:
```diff
@@ -0,0 +1,254 @@
+    }
+}
+
+#[test]
```
these test functions should also be inside a `mod test`.

ramsayleung commented:
I have refactored `enums` module with `strum` crate, reducing line-of-code from 7396 to 6972, generated by `tokei`. It's much clear and readable than before, but we have introduced a new dependency, I think about that will it increase the compilation time ? The number of compilation units increased from 159 to 163 with default feature.

marioortizmanero commented:
That's great! Have you measured the compilation times, rather than the number of units?

ramsayleung commented:
Yes, I have measured the compilation times:

```sh
git checkout 72c6ee3 # the last commit without `strum` crate.
> hyperfine --warmup 2 'cargo clean && cargo build --no-default-features --features client-reqwest' -i
Benchmark #1: cargo clean && cargo build --no-default-features --features client-reqwest
  Time (mean ± σ):     86.715 s ±  3.165 s    [User: 276.518 s, System: 14.973 s]
  Range (min … max):   82.019 s … 91.968 s    10 runs

git checkout ramsay_clean_up_rspotify 
> hyperfine --warmup 2 'cargo clean && cargo build --no-default-features --features client-reqwest' -i
Benchmark #1: cargo clean && cargo build --no-default-features --features client-reqwest
  Time (mean ± σ):     88.646 s ±  1.688 s    [User: 285.274 s, System: 15.676 s]
  Range (min … max):   84.932 s … 90.441 s    10 runs
```

marioortizmanero commented on src/model/enums/misc.rs:
```diff
@@ -0,0 +1,105 @@
+
+/// disallow: interrupting_playback, pausing, resuming, seeking, skipping_next, skipping_prev, toggling_repeat_context, toggling_shuffle, toggling_repeat_track, transferring_playback
+#[derive(
+    Clone, Serialize, Deserialize, Copy, PartialEq, Eq, Debug, Hash, EnumString, AsRefStr, Display,
```
May I ask why we use `AsRefStr` instead of `ToString` for these derive macros? What is the `disallow: ...` part in the comment as well?

marioortizmanero commented on src/model/enums/misc.rs:
```diff
@@ -0,0 +1,105 @@
+use serde::{Deserialize, Serialize};
+use strum::{AsRefStr, Display, EnumString};
+
+/// disallow: interrupting_playback, pausing, resuming, seeking, skipping_next, skipping_prev, toggling_repeat_context, toggling_shuffle, toggling_repeat_track, transferring_playback
```
What is the `disallow: ...` part in the comment?

marioortizmanero commented on src/model/enums/misc.rs:
```diff
@@ -0,0 +1,105 @@
+use serde::{Deserialize, Serialize};
+use strum::{AsRefStr, Display, EnumString};
+
+/// disallow: interrupting_playback, pausing, resuming, seeking, skipping_next, skipping_prev, toggling_repeat_context, toggling_shuffle, toggling_repeat_track, transferring_playback
```
What is the `disallow: ...` part in the comment?

ramsayleung replied:
｀disallow｀ just means `disallow key:`, I would rename it to `disallow key`.

marioortizmanero replied:
I still don't quite understand it, sorry. What's that for?

ramsayleung replied:
Disallow means the [disallows object](https://developer.spotify.com/documentation/web-api/reference/object-model/#disallows-object), which is used in the endpoint [`get-users-currently-playing-track`](https://developer.spotify.com/documentation/web-api/reference/player/get-the-users-currently-playing-track/), is is clear to you now?

marioortizmanero replied:
Ahhh okay, yes, sorry. The link to the reference helped. For some reason I thought it was something for the linter.

Can you add the references for these structs/enums like the endpoints? They are usually like this:

```rust
/// Endpoint: description
///
/// ...
///
/// [Reference](reference)
```

just for consistency

ramsayleung replied:
Yep, I have already updated this. This comments is outdated.

marioortizmanero replied:
Great. Can you make it 80 characters long as always?

ramsayleung replied:
Yes, of course. I think it's a better way to let `rustfmt` handle it :)

marioortizmanero replied:
rustfmt doesn't modify comments AFAIK

marioortizmanero commented on src/model/enums/misc.rs:
```diff
@@ -0,0 +1,105 @@
+
+/// disallow: interrupting_playback, pausing, resuming, seeking, skipping_next, skipping_prev, toggling_repeat_context, toggling_shuffle, toggling_repeat_track, transferring_playback
+#[derive(
+    Clone, Serialize, Deserialize, Copy, PartialEq, Eq, Debug, Hash, EnumString, AsRefStr, Display,
```
May I ask why we use `AsRefStr` instead of `ToString` for these derive macros?

ramsayleung replied:
Actually, I can't tell the difference between `EnumString` and `Tostring`, which are almost same, so I choose `EnumString` with additional attribute `#[strum(serialize_all = "snake_case")]`.

`AsRefStr` is another trait to derive, which performs as same as the old function `as_str()`. Thus, instead of prefer `Tostring` to `AsRefStr`, I prefer to keep them both, but I choose `EnumString` instead of `ToString`.

marioortizmanero replied:
The difference is that `EnumString` implements `FromStr`, and `ToString` implents `ToString`. So you can't just choose between these two because they do different things. If anything, the choice is between `ToString`, which returns a `String`, and `AsRefStr`, which returns a `&str`. But IMO the former makes more sense in the context of what we need: to convert something into a string.

But that makes me wonder what these derives are even for. `aspotify` for example doesn't seem to implement `ToString` and `FromStr`. Are they needed for serialization or something? Isn't that done by `serde`?

ramsayleung replied:
>  Isn't that done by serde?

No, it's nothing related to `serialization`, `to_string()` is for setting something like a query parameter:
```rust
    #[maybe_async]
    pub async fn tracks<'a>(
        &self,
        track_ids: impl IntoIterator<Item = &'a str>,
        market: Option<Country>,
    ) -> ClientResult<FullTracks> {
        // TODO: this can be improved
        let mut ids: Vec<String> = vec![];
        for track_id in track_ids {
            ids.push(self.get_id(Type::Track, track_id));
        }

        let mut params = Query::new();
        if let Some(market) = market {
            params.insert("market".to_owned(), market.to_string());
        }

        let url = format!("tracks/?ids={}", ids.join(","));
        let result = self.get(&url, None, &params).await?;
        self.convert_result(&result)
    }

```
`enums` is not only for model, but also used in somewhere else.

marioortizmanero replied:
> No, it's nothing related to serialization, to_string() is for setting something like a query parameter:

OK, thanks for clearing it up. So wouldn't it make more sense to convert to a string with `.to_string()` in that case instead of `.as_ref()`? AFAIK the latter is mostly used in smart pointers like `Box`.

ramsayleung replied:
> AFAIK the latter is mostly used in smart pointers like Box. Some of your enums derive both AsRefStr and Display; I think we would be fine with just a single ToString.

Yep, good point, I would replace all `EnumString, AsRefStr, Display` to `ToString`, your point makes sense :)

marioortizmanero replied:
I've been investigating and in fact, we could avoid `strum` altogether and use this from `serde_json` - https://docs.serde.rs/serde_json/ser/fn.to_string.html

marioortizmanero replied:
Great! That should reduce compile times as well.

marioortizmanero replied:
All I've found is this crate - https://docs.rs/serde_plain/0.3.0/serde_plain/fn.to_string.html - but I'm not sure if it's better than using `strum`. It saves us from repeating the lowercase conversions and renames, and it seems to be lighter than `strum`, but you'd have to use `serde_plain::to_string(&val)?` instead of `var.to_string()`.

ramsayleung replied:
IMHO, this syntax is a little bit weird, perhaps we could find another way to do so.
 
> you'd have to use serde_plain::to_string(&val)? instead of var.to_string().

marioortizmanero replied:
I think it's impossible:

1. `Serialize` is an external trait from `serde`
2. `ToString` is an external trait from `std`

Thus, a crate like `serde_json` can't have a blanket implementation of `ToString` for all `T: Serialize` types because both are external to it. If anything, it could create an equivalent trait for `ToString`, like `OurToString`, but that wouldn't play well with generic types over `ToString`.

ramsayleung replied:
Yeah, I agree with you, `strum` is a solid solution for our case. By the way, I have pushed some new commits, including a bit model modification. I will keep refactoring model after this PR merged.

marioortizmanero commented on src/model/enums/country.rs:
```diff
@@ -0,0 +1,525 @@
+#[derive(
+    Clone, Serialize, Deserialize, Copy, PartialEq, Eq, Debug, EnumString, AsRefStr, Display,
+)]
+pub enum Country {
```
Isn't there a crate for this already? I would really prefer that if possible. It saves us 500 lines of untested and unmaintained code, compared to a well known crate. `aspotify` seems to use `isocode`: https://docs.rs/aspotify/0.6.1/aspotify/enum.CountryCode.html

ramsayleung replied:
I don't think it's necessary to add a new crate just to take one `enum` from this crate, which will break the consistency of `enums` module, some enums are defined in `enums` module, and others are defined in another crate, adding such a new crate doesn't pay off.

marioortizmanero commented on src/model/enums/types.rs:
```diff
@@ -0,0 +1,155 @@
+pub enum CurrentlyPlayingType {
+    Track,
+    Episode,
+    #[strum(serialize = "ad")]
```
Why is this done with `strum` and not `serde`?

ramsayleung replied:
Yes, it's a problem, I'll fix it.

marioortizmanero commented on src/model/enums/mod.rs:
```diff
@@ -0,0 +1,18 @@
+
+pub use country::Country;
+
+pub use misc::DisallowKey;
```
Can these be grouped up with `use thing::{a,b,c}`?

ramsayleung replied:
Sure!

marioortizmanero commented on src/model/enums/mod.rs:
```diff
@@ -0,0 +1,18 @@
+pub use misc::RepeatState;
+pub use misc::TimeRange;
+
+pub use types::AdditionalType;
```
Same as above

marioortizmanero reviewed (approved):
I think this is ready to be merged.

ramsayleung commented:
Merged :) 

I think I will take times to cross off more items in #127 :)
